### PR TITLE
Activity Log: Linearize and page view based on perishable filters

### PR DESCRIPTION
**Pull request by dmsnell**
This is a dramatic change to the Activity Log display - it removes the monthly/daily list view and linearizes it instead. It removes complicated persistence going on to maintain internal state when updating the events query.

@MichaelArestad if you look at the live branch you will see I have some bad date headers. I think we need to make some decisions on how to change the way we show time. _if_ we always started at today it would be clearer but _if_ we are going to filter and search we won't have that liberty. for now I was tossing around the idea of putting a real header for the month and then put the day of the month somehow on each event next to the hour-time

**Current status**
 - Cannot change the query parameters, this just shows up to 1000 most-recent events
 - Stopped polling for new events. Probably will add that back in before this PR is done but I don't want to append new events to the list. Instead I wanted to show a kind of indicator like the "orange pill" in the reader view. Would be a good follow-up PR to this one in order to keep scope down.

**Need help**
 - <del>We are putting in a header for every month-change and on the first event (top of the list) but I'm no good at styling so help is appreciated.</del>

**In Master**
monthlyview mov

**In this branch**
linearview2 mov

### CODE REVIEW

File: client/my-sites/stats/activity-log/index.jsx [hunk 1]
```diff
@@ -472,6 +325,38 @@ class ActivityLog extends Component {
...
+
+			return ( { rewindId } ) => {
+				const ts = 1000 * rewindId;
+
+				if ( null === last ) {
+					last = ts;
+					return (
+						<p key="time-period-head">
+							{ moment( ts ).format(
+								translate( 'MMMM Y', { context: 'moment.js format string to show month and year' } )
```
Comment (a8ci18n) [inline client/my-sites/stats/activity-log/index.jsx]: :information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

**Comment by beaulebens** [comment]
2 quick notes after I tried this out on calypso.live.

1. If you're looking for a design, check out the internal build

**Comment by sirreal** [comment]
Something that jumped out at me in calypso.live was that the pagination felt backwards:

pagination

That may be because I'm used the the monthly nav, which went to the left to go back. Should we consider reversing the pagination?

**Comment by dmsnell** [comment]
> For "paging", it might work better (although be significantly more complicated :) ) to somehow provide time-based anchors to the specific event at the top of the "stream", and update that as you infinite-scroll through time. Just a thought though, don't really know how that'd feel/how hard it'd be to make it make sense.

I'm going to strongly urge us to give up infinite scroll, especially for now. It adds considerable UI complexity and provides very little value. What value does it even bring? If we're thinking about filtering this view then I don't see any benefit other than providing extra hours we can spend in dev time :wink:

Personally _I_ don't want to be scrolling through hundreds of events to find what I'm looking for. I'd rather narrow my search requirements.

**Comment by dmsnell** [comment]
> That may be because I'm used the the monthly nav, which went to the left to go back. Should we consider reversing the pagination?

Maybe but I don't think so. @MichaelArestad and @keoshi can chime in. the paging is borrowed from Comments Management and I think reversing the paging would end up needlessly confusing. What other paging strip goes in reverse? I'd rather have incredibly-strong justification that it's better to do than to break a seemingly ubiquitous idiom.

**Comment by keoshi** [comment]
> per-day headers

This is what @beaulebens was referring to, I think it can work much better than adding the day on the `activity-log-item__time` (from p9c7Jx-nz-p2 ):

list-view

I think I have bandwidth today to work on a quick mock for this.

> pagination felt backwards

I have to agree. It feels especially weird here since the timeline pattern is the foundation for the entire interface. What confuses me the most is clicking on “Previous” to see what actually are “Latest” items.

I also think this is pretty much a personal preference and just different frame of mind, since I've seen WP themes treat “previous/next” and “older/newer” is complete opposite ways. Just want us to be mindful that this tool isn't isolated from the rest of Calypso. We've got this pattern in comments, but Activity is also very much a part of stats (for right now, at least) and the pattern there is based on a calendar paradigm where in LTR languages means “left to go back”.

**Comment by dmsnell** [comment]
> I also think this is pretty much a personal preference and just different frame of mind, since I've seen WP themes (for example) treat “previous/next” and “older/newer” is complete opposite ways. Just want us to be mindful that this tool isn't isolated from the rest of Calypso. We've got this pattern in comments, but Activity is also very much a part of stats (for right now, at least) and the pattern there is based on a calendar paradigm where in LTR languages means “left to go back”.

my point is more a result of changing our paradigm from a calendar view to a filtered view. we're paging through results, not through time primarily. I'd be fine using `older` and `newer` rather than `prev`/`next`.

note how in `stats` we're actually panning the view left and right corresponding to the arrows. in activity log we're going from one page in the results to another and it's not required to go from one to the next - it can skip results.

---

alternatively not paging is an option but I'd advise strongly against that. we could just display the 1000 results in a single view and then the paging goes away. however, by paging in JS it's quite simple in the code and easy on the view. again, who wants to scan through 500 events? if it's not in the first page or two refine the search

**Comment by dmsnell** [comment]
> I think it can work much better than adding the day on the activity-log-item__time

my biggest concern there is that we end up quite frequently where every single event gets its own date header, at which point we've effectively added the date to every event itself.

even just browsing the log by itself as-is we have plenty of days with a single event in them, but as soon as we expand filtering we'll have even more since the act of filtering narrows the number of events on any given day that will appear.

in other words, if we really want to add the header for every event, let's do it. we've done crazier things before.

**Comment by keoshi** [comment]
> using `older` and `newer` rather than `prev`/`next`

That sounds good to me!

> not paging is an option

I don't think anyone is pushing for that, and to me that's a non-started if we want to make this usable.

> my biggest concern there is that we end up quite frequently where every single event gets its own date header

Agreed, this is one of those we have to try and click around to get a feel for what works best. I'm going to play around with it and see what could work.

**Comment by samhotchkiss** [comment]
This is super exciting and a step in the right direction, for sure!

A few things:

1. I want to make sure that we're thinking about how we could link to individual events in the timeline.  This is important for happiness, and I could see if being useful to us in the future, as well.
2. There should definitely be more events per page, right?  Is there any reason not to do 100 events/page?

As far as the date declarations go, I was imagining something like this:
image

This, IMO, wouldn't get too out of hand if you only had one event per day.  WAY better than having to expand each day individually.

**Comment by dmsnell** [comment]
Splitting out the `newer`/`older` changes into #23989

**Comment by dmsnell** [comment]
> I want to make sure that we're thinking about how we could link to individual events in the timeline. This is important for happiness, and I could see if being useful to us in the future, as well.

@samhotchkiss thankfully the fundamental change in this PR is that we can now pinpoint which events we want to show. "linking" to an event would end up being nothing more than setting the query to pull a specific activity id. I'd like to do that in a separate PR once this merges in order to keep the scope limited here. although we're not adding too many lines, we're removing many and changing lots of behaviors as it is. thanks for keeping that goal in the open though!

**Comment by dmsnell** [comment]
> There should definitely be more events per page, right? Is there any reason not to do 100 events/page?

Definitely! I'm using 10 just for the development work so that I can more quickly see the entire page and guarantee I have plenty of pages to navigate through. Probably 100 is reasonable once we get there, or should we decided we can even expose a dropdown as it in `wp-admin` for things like the post list.

**Comment by dmsnell** [comment]
Custom pagination labels merged so now this shows `newer`/`older`

**Comment by MichaelArestad** [comment]
Updated the styles:
* changed time period formatting and styles
* moved pagination out of the wrapper to prevent the timeline "line" from stretching past it
* added a gradient to the end of the timeline "line"

It should look like this, now:

image

image

image

**Comment by keoshi** [comment]
Looking good, @MichaelArestad !

**Comment by samhotchkiss** [comment]
@dmsnell: Can you go ahead and up the per-page count to 100 so we can see how that feels?

**Comment by dmsnell** [comment]
> Can you go ahead and up the per-page count to 100 so we can see how that feels?

good call @samhotchkiss - done!

**Comment by beaulebens** [comment]
This is feeling _so much better_!

One thing I noticed with the new layout is that the loading placeholders are "off" now, since they are based on the previous daily cards. It's hard to capture them because they disappear as soon as the events are loaded, but here's what I managed to get:

screen shot 2018-04-11 at 11 54 24 am

We could probably blank out a couple of more-like-the-normal-event displays something like this as a replacement (and set avatar to mystery man, or else a Jetpack logo perhaps):

screen shot 2018-04-11 at 12 00 23 pm

cc @keoshi for a better/real approach to this, vs my frankenhacking.

**Comment by keoshi** [comment]
@MichaelArestad as discussed in today's call, here's my attempt at making the headers take less vertical space if you want to tweak that:

```
.activity-log__time-period {
    margin: 8px 0;
    padding: 6px 0 4px;
}
```

image

**Comment by dmsnell** [comment]
> the other is a glitch when switching sites: I was in the page 3 of the AL of a site, and switched to this site that didn't have page 3. It looked like this

thanks @eliorivero - fixed that in 988e177 so that the current page resets to 1 on site transitions

**Comment by dmsnell** [comment]
> invalid date that others have found

after some investigation I found that the `rewind_id` was an empty string and probably from a very old and invalid event. I don't want to block this PR on that note as all new and recent events (within the past several months) should have valid `rewind_id`s.

in the meantime I'm going to update the API so that it prunes out these events. thanks for reporting it!
